client/keys/add.go

```diff
@@ -355,68 +361,72 @@ func AddNewKeyRequestHandler(indent bool) http.HandlerFunc {
 		var m NewKeyBody
 
 		kb, err := GetKeyBaseWithWritePerm()
-		if err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			w.Write([]byte(err.Error()))
+		if checkErrorREST(w, http.StatusInternalServerError, err) {
 			return
 		}
 
 		body, err := ioutil.ReadAll(r.Body)
-		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			w.Write([]byte(err.Error()))
+		if checkErrorREST(w, http.StatusBadRequest, err) {
 			return
 		}
+
 		err = json.Unmarshal(body, &m)
-		if err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			w.Write([]byte(err.Error()))
+		if checkErrorREST(w, http.StatusBadRequest, err) {
 			return
 		}
+
+		// Check parameters
 		if m.Name == "" {
-			w.WriteHeader(http.StatusBadRequest)
-			err = errMissingName()
-			w.Write([]byte(err.Error()))
+			checkErrorREST(w, http.StatusBadRequest, errMissingName())
 			return
 		}
 		if m.Password == "" {
-			w.WriteHeader(http.StatusBadRequest)
-			err = errMissingPassword()
-			w.Write([]byte(err.Error()))
+			checkErrorREST(w, http.StatusBadRequest, errMissingPassword())
+			return
+		}
+
+		mnemonic := m.Mnemonic
+		// if mnemonic is empty, generate one
+		if mnemonic == "" {
+			mnemonic = generateMnemonic(keys.Secp256k1)
+		}
+		if !bip39.IsMnemonicValid(mnemonic) {
+			checkErrorREST(w, http.StatusBadRequest, errInvalidMnemonic())
+		}
+
+		if m.Account < 0 || m.Account > maxValidAccountValue {
+			checkErrorREST(w, http.StatusBadRequest, errInvalidAccountNumber())
+			return
+		}
+
+		if m.Index < 0 || m.Index > maxValidIndexalue {
+			checkErrorREST(w, http.StatusBadRequest, errInvalidIndexNumber())
 			return
 		}
 
 		// check if already exists
-		infos, err := kb.List()
```

Comment:
I don't really get why this code was iterating over the list of all keys... wouldn't it be simpler error-check `kb.Get(m.Name)`?

Reply:
yes, didn't want to rewrite too much code. Done.